flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java

```diff
@@ -67,361 +89,422 @@
-
-					} finally {
-						// close and prepare for the next iteration
-						synchronized (checkpointLock) {
-							this.format.close();
-							this.isSplitOpen = false;
-							this.currentSplit = null;
-						}
-					}
+		switch (state) {
+			case IDLE:
+				switchState(ReaderState.CLOSED);
+				break;
+			case CLOSED:
+				LOG.warn("operator is already closed, doing nothing");
+				return;
+			default:
+				switchState(ReaderState.CLOSING);
+				while (!state.isTerminal()) {
+					executor.yield();
 				}
+		}
 
-			} catch (Throwable e) {
-
-				getContainingTask().handleAsyncException("Caught exception when processing split: " + currentSplit, e);
+		try {
+			sourceContext.emitWatermark(Watermark.MAX_WATERMARK);
+		} catch (Exception e) {
+			LOG.warn("unable to emit watermark while closing", e);
+		}
 
-			} finally {
-				synchronized (checkpointLock) {
-					LOG.debug("Reader terminated, and exiting...");
+		cleanUp();
+	}
 
-					try {
-						this.format.closeInputFormat();
-					} catch (IOException e) {
-						getContainingTask().handleAsyncException(
-							"Caught exception from " + this.format.getClass().getName() + ".closeInputFormat() : " + e.getMessage(), e);
-					}
-					this.isSplitOpen = false;
-					this.currentSplit = null;
-					this.isRunning = false;
+	private void cleanUp() {
+		LOG.debug("cleanup, state={}", state);
 
-					checkpointLock.notifyAll();
-				}
-			}
-		}
+		RunnableWithException[] runClose = {
+				() -> sourceContext.close(),
+				() -> output.close(),
+				() -> format.close(),
+				() -> format.closeInputFormat()};
+		Exception ef = null;
 
-		private List<TimestampedFileInputSplit> getReaderState() throws IOException {
-			List<TimestampedFileInputSplit> snapshot = new ArrayList<>(this.pendingSplits.size());
-			if (currentSplit != null) {
-				if (this.format instanceof CheckpointableInputFormat && this.isSplitOpen) {
-					Serializable formatState =
-						((CheckpointableInputFormat<TimestampedFileInputSplit, Serializable>) this.format).getCurrentState();
-					this.currentSplit.setSplitState(formatState);
-				}
-				snapshot.add(this.currentSplit);
+		for (RunnableWithException r : runClose) {
+			try {
+				r.run();
+			} catch (Exception e) {
+				ef = ExceptionUtils.firstOrSuppressed(ef, e);
 			}
-			snapshot.addAll(this.pendingSplits);
-			return snapshot;
 		}
-
-		public void cancel() {
-			this.isRunning = false;
-		}
-
-		public void close() {
-			this.shouldClose = true;
+		currentSplit = null;
+		if (ef != null) {
+			throw new FlinkRuntimeException("Unable to properly cleanup ContinuousFileReaderOperator", ef);
 		}
 	}
 
-	//	---------------------			Checkpointing			--------------------------
-
 	@Override
 	public void snapshotState(StateSnapshotContext context) throws Exception {
 		super.snapshotState(context);
 
 		checkState(checkpointedState != null,
-			"The operator state has not been properly initialized.");
+				"The operator state has not been properly initialized.");
```

Comment:
nit: 
1. incorrect formatting (and in couple of lines below).
2. please try to not mix formatting (and refactoring) changes with functional changes